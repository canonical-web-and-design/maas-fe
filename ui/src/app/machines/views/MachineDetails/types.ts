import type { HardwareType } from "app/base/enum";
import type { MachineAction } from "app/store/general/types";
import type { Script } from "app/store/script/types";

export type SelectedAction = {
  name: MachineAction["name"];
  sentence?: MachineAction["sentence"];
  formProps?: {
    hardwareType?: HardwareType;
    applyConfiguredNetworking?: Script["apply_configured_networking"];
  };
};

export type SetSelectedAction = (
  action: SelectedAction | null,
  deselect?: boolean
) => void;
