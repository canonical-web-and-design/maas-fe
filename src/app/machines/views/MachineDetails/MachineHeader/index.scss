@mixin MachineHeader {

  // These media queries are for when the screen is too small to display the action group without wrapping.
  // Wrapping begins before $breakpoint-small, hence the direct reference to pixel width.
  @media screen and (max-width: ($breakpoint-node-action-menu-group - 1)) {
    .p-node-action-menu-group {
      display: none;
    }
  }

  @media screen and (min-width: $breakpoint-node-action-menu-group) {
    .p-action-menu {
      display: none;
    }
  }
}