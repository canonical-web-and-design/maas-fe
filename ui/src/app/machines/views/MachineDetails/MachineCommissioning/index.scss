@mixin MachineCommissioning {
  .p-table--machine-commissioning {
    th,
    td {
      &:nth-child(1) {
        min-width: 30%;
      }
      &:nth-child(2) {
        min-width: 15%;
      }
      &:nth-child(3) {
        min-width: 10%;
      }
      &:nth-child(4) {
        min-width: 20%;
      }
      &:nth-child(5) {
        min-width: 15%;
      }
      &:nth-child(6) {
        min-width: 10%;
      }
    }
  }
}
