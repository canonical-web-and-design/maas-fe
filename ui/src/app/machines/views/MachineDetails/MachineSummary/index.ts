export { default } from "./MachineSummary";
export type { SelectedAction, SetSelectedAction } from "./MachineSummary";
