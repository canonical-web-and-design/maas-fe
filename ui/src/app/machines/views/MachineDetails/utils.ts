import { scriptStatus } from "app/base/enum";
import type { ScriptResult } from "app/store/scriptresult/types";

export const getTestResultsIcon = (result: ScriptResult): string => {
  switch (result.status) {
    case scriptStatus.PENDING:
      return "p-icon--pending";
    case scriptStatus.RUNNING:
    case scriptStatus.APPLYING_NETCONF:
    case scriptStatus.INSTALLING:
      return "p-icon--running";
    case scriptStatus.PASSED:
      return "p-icon--success";
    case scriptStatus.FAILED:
    case scriptStatus.ABORTED:
    case scriptStatus.DEGRADED:
    case scriptStatus.FAILED_APPLYING_NETCONF:
    case scriptStatus.FAILED_INSTALLING:
      return "p-icon--error";
    case scriptStatus.TIMEDOUT:
      return "p-icon--timed-out";
    case scriptStatus.SKIPPED:
      return "p-icon--warning";
    case scriptStatus.NONE:
      return "";
    default:
      return "p-icon--help";
  }
};
