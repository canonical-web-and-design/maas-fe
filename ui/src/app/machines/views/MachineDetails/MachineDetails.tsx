import { Redirect, Route, Switch } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import { useParams } from "react-router";
import React, { useEffect, useState } from "react";

import type { RouteParams } from "app/base/types";
import { machine as machineActions } from "app/base/actions";
import MachineHeader from "./MachineHeader";
import MachineNotifications from "./MachineNotifications";
import machineSelectors from "app/store/machine/selectors";
import MachineStorage from "./MachineStorage";
import MachineSummary, { SelectedAction } from "./MachineSummary";
import Section from "app/base/components/Section";
import type { RootState } from "app/store/root/types";

const MachineDetails = (): JSX.Element => {
  const dispatch = useDispatch();
  const params = useParams<RouteParams>();
  const { id } = params;
  const machine = useSelector((state: RootState) =>
    machineSelectors.getById(state, id)
  );
  const machinesLoaded = useSelector(machineSelectors.loaded);
  const [selectedAction, setSelectedAction] = useState<SelectedAction | null>(
    null
  );

  useEffect(() => {
    dispatch(machineActions.get(id));
    // Set machine as active to ensure all machine data is sent from the server.
    dispatch(machineActions.setActive(id));

    // Unset active machine on cleanup.
    return () => {
      dispatch(machineActions.setActive(null));
    };
  }, [dispatch, id]);

  // If machine has been deleted, redirect to machine list.
  if (machinesLoaded && !machine) {
    return <Redirect to="/machines" />;
  }

  return (
    <Section
      header={
        <MachineHeader
          selectedAction={selectedAction}
          setSelectedAction={setSelectedAction}
        />
      }
      headerClassName="u-no-padding--bottom"
    >
      <MachineNotifications />
      {machine && (
        <Switch>
          <Route exact path="/machine/:id/summary">
            <MachineSummary setSelectedAction={setSelectedAction} />
          </Route>
          <Route exact path="/machine/:id/storage">
            <MachineStorage />
          </Route>
          <Route exact path="/machine/:id">
            <Redirect to={`/machine/${id}/summary`} />
          </Route>
        </Switch>
      )}
    </Section>
  );
};

export default MachineDetails;
