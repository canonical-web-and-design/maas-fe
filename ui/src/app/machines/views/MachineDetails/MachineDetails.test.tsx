import { mount } from "enzyme";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import React from "react";

import {
  machine as machineFactory,
  machineState as machineStateFactory,
  rootState as rootStateFactory,
} from "testing/factories";
import MachineDetails from "./MachineDetails";
import type { RootState } from "app/store/root/types";

const mockStore = configureStore();

describe("MachineDetails", () => {
  let state: RootState;
  beforeEach(() => {
    state = rootStateFactory({
      machine: machineStateFactory({
        items: [machineFactory({ system_id: "abc123" })],
      }),
    });
  });

  it("renders", () => {
    const store = mockStore(state);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[{ pathname: "/machine/abc123", key: "testKey" }]}
        >
          <MachineDetails />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.find("MachineDetails")).toMatchSnapshot();
  });
});
