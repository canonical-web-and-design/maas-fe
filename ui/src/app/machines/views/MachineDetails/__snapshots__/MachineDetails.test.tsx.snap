// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MachineDetails renders 1`] = `
<MachineDetails>
  <Section
    header={<MachineHeader />}
    headerClassName="u-no-padding--bottom"
  >
    <div
      className="section"
    >
      <Strip
        className="section__header u-no-padding--bottom"
        element="header"
        shallow={true}
      >
        <header
          className="section__header u-no-padding--bottom p-strip is-shallow"
        >
          <Row>
            <div
              className="row"
            >
              <Col
                size="12"
              >
                <div
                  className="col-12"
                >
                  <MachineHeader>
                    <SectionHeader
                      loading={true}
                      title=""
                    >
                      <div
                        className="u-flex--between u-flex--wrap"
                      >
                        <ul
                          className="p-inline-list"
                        >
                          <li
                            className="p-inline-list__item p-heading--four"
                            data-test="section-header-title"
                          />
                          <li
                            className="p-inline-list__item last-item u-text--light"
                          >
                            <Spinner
                              className="u-no-padding u-no-margin"
                              inline={true}
                              text="Loading..."
                            >
                              <span
                                className="u-no-padding u-no-margin p-spinner p-text--default p-spinner--inline"
                              >
                                <i
                                  className="p-icon--spinner u-animation--spin"
                                />
                                <span
                                  className="p-icon__text"
                                >
                                  Loading...
                                </span>
                              </span>
                            </Spinner>
                          </li>
                        </ul>
                      </div>
                    </SectionHeader>
                  </MachineHeader>
                </div>
              </Col>
            </div>
          </Row>
        </header>
      </Strip>
      <Strip
        element="main"
        includeCol={false}
        rowClassName="section__content-wrapper"
        shallow={true}
      >
        <main
          className="p-strip is-shallow"
        >
          <Row
            className="section__content-wrapper"
          >
            <div
              className="section__content-wrapper row"
            >
              <Col
                className="section__content"
                size={12}
              >
                <div
                  className="section__content col-12"
                >
                  <NotificationList>
                    <NotificationGroup
                      key="caution"
                      notifications={
                        Array [
                          Object {
                            "admins": true,
                            "category": "warning",
                            "created": "Wed, 08 Jul. 2020 05:35:4",
                            "dismissable": true,
                            "id": 11,
                            "ident": null,
                            "message": "Testing notification",
                            "updated": "Wed, 08 Jul. 2020 05:35:4",
                            "user": Object {
                              "completed_intro": true,
                              "email": "email1@example.com",
                              "global_permissions": Array [
                                "machine_create",
                              ],
                              "id": 12,
                              "is_superuser": true,
                              "last_name": "Full Name jr.",
                              "sshkeys_count": 3,
                              "username": "user1",
                            },
                            "users": true,
                          },
                          Object {
                            "admins": true,
                            "category": "warning",
                            "created": "Wed, 08 Jul. 2020 05:35:4",
                            "dismissable": true,
                            "id": 13,
                            "ident": null,
                            "message": "Testing notification",
                            "updated": "Wed, 08 Jul. 2020 05:35:4",
                            "user": Object {
                              "completed_intro": true,
                              "email": "email2@example.com",
                              "global_permissions": Array [
                                "machine_create",
                              ],
                              "id": 14,
                              "is_superuser": true,
                              "last_name": "Full Name jr.",
                              "sshkeys_count": 3,
                              "username": "user2",
                            },
                            "users": true,
                          },
                          Object {
                            "admins": true,
                            "category": "warning",
                            "created": "Wed, 08 Jul. 2020 05:35:4",
                            "dismissable": true,
                            "id": 15,
                            "ident": null,
                            "message": "Testing notification",
                            "updated": "Wed, 08 Jul. 2020 05:35:4",
                            "user": Object {
                              "completed_intro": true,
                              "email": "email3@example.com",
                              "global_permissions": Array [
                                "machine_create",
                              ],
                              "id": 16,
                              "is_superuser": true,
                              "last_name": "Full Name jr.",
                              "sshkeys_count": 3,
                              "username": "user3",
                            },
                            "users": true,
                          },
                          Object {
                            "admins": true,
                            "category": "warning",
                            "created": "Wed, 08 Jul. 2020 05:35:4",
                            "dismissable": true,
                            "id": 17,
                            "ident": null,
                            "message": "Testing notification",
                            "updated": "Wed, 08 Jul. 2020 05:35:4",
                            "user": Object {
                              "completed_intro": true,
                              "email": "email4@example.com",
                              "global_permissions": Array [
                                "machine_create",
                              ],
                              "id": 18,
                              "is_superuser": true,
                              "last_name": "Full Name jr.",
                              "sshkeys_count": 3,
                              "username": "user4",
                            },
                            "users": true,
                          },
                          Object {
                            "admins": true,
                            "category": "warning",
                            "created": "Wed, 08 Jul. 2020 05:35:4",
                            "dismissable": true,
                            "id": 19,
                            "ident": null,
                            "message": "Testing notification",
                            "updated": "Wed, 08 Jul. 2020 05:35:4",
                            "user": Object {
                              "completed_intro": true,
                              "email": "email5@example.com",
                              "global_permissions": Array [
                                "machine_create",
                              ],
                              "id": 20,
                              "is_superuser": true,
                              "last_name": "Full Name jr.",
                              "sshkeys_count": 3,
                              "username": "user5",
                            },
                            "users": true,
                          },
                        ]
                      }
                      type="caution"
                    >
                      <div
                        className="p-notification--group"
                      >
                        <Notification
                          type="caution"
                        >
                          <div
                            className="p-notification--caution"
                          >
                            <p
                              className="p-notification__response"
                            >
                              <Button
                                appearance="link"
                                aria-label="5 caution, click to open messages."
                                onClick={[Function]}
                              >
                                <button
                                  aria-label="5 caution, click to open messages."
                                  className="p-button--link"
                                  onClick={[Function]}
                                >
                                  <span
                                    className="p-notification__status"
                                    data-test="notification-count"
                                  >
                                    5 Warnings
                                  </span>
                                  <small>
                                    <i
                                      className="p-icon--expand"
                                    />
                                  </small>
                                </button>
                              </Button>
                              <Button
                                appearance="link"
                                className="p-notification__action u-nudge-right"
                                onClick={[Function]}
                              >
                                <button
                                  className="p-button--link p-notification__action u-nudge-right"
                                  onClick={[Function]}
                                >
                                  Dismiss all
                                </button>
                              </Button>
                            </p>
                          </div>
                        </Notification>
                      </div>
                    </NotificationGroup>
                    <Notification
                      close={[Function]}
                      data-test="message"
                      key="6"
                      timeout={5000}
                      type="caution"
                    >
                      <div
                        className="p-notification--caution"
                        data-test="message"
                      >
                        <p
                          className="p-notification__response"
                        >
                          Test message
                        </p>
                        <button
                          aria-label="Close notification"
                          className="p-icon--close"
                          onClick={[Function]}
                        >
                          Close
                        </button>
                      </div>
                    </Notification>
                    <Notification
                      close={[Function]}
                      data-test="message"
                      key="7"
                      timeout={5000}
                      type="caution"
                    >
                      <div
                        className="p-notification--caution"
                        data-test="message"
                      >
                        <p
                          className="p-notification__response"
                        >
                          Test message
                        </p>
                        <button
                          aria-label="Close notification"
                          className="p-icon--close"
                          onClick={[Function]}
                        >
                          Close
                        </button>
                      </div>
                    </Notification>
                    <Notification
                      close={[Function]}
                      data-test="message"
                      key="8"
                      timeout={5000}
                      type="caution"
                    >
                      <div
                        className="p-notification--caution"
                        data-test="message"
                      >
                        <p
                          className="p-notification__response"
                        >
                          Test message
                        </p>
                        <button
                          aria-label="Close notification"
                          className="p-icon--close"
                          onClick={[Function]}
                        >
                          Close
                        </button>
                      </div>
                    </Notification>
                    <Notification
                      close={[Function]}
                      data-test="message"
                      key="9"
                      timeout={5000}
                      type="caution"
                    >
                      <div
                        className="p-notification--caution"
                        data-test="message"
                      >
                        <p
                          className="p-notification__response"
                        >
                          Test message
                        </p>
                        <button
                          aria-label="Close notification"
                          className="p-icon--close"
                          onClick={[Function]}
                        >
                          Close
                        </button>
                      </div>
                    </Notification>
                    <Notification
                      close={[Function]}
                      data-test="message"
                      key="10"
                      timeout={5000}
                      type="caution"
                    >
                      <div
                        className="p-notification--caution"
                        data-test="message"
                      >
                        <p
                          className="p-notification__response"
                        >
                          Test message
                        </p>
                        <button
                          aria-label="Close notification"
                          className="p-icon--close"
                          onClick={[Function]}
                        >
                          Close
                        </button>
                      </div>
                    </Notification>
                  </NotificationList>
                </div>
              </Col>
            </div>
          </Row>
        </main>
      </Strip>
    </div>
  </Section>
</MachineDetails>
`;
